[TF] Set default RPATH to toolchain stdlib instead of `/usr/lib/swift` on macOS.

dan-zheng opened the pull request:
Previously, the default RPATH was set to `/usr/lib/swift`.
This caused linker issues on macOS, as tensorflow-branch-specific modules like
TensorFlow and Python do not exist in `/usr/lib/swift`.

This patch defaults the `-toolchain-stdlib-rpath` flag to be true, so the
default RPATH is the toolchain standard library instead of `/usr/lib/swift`.

With this default, the linker issues are fixed. A `-no-toolchain-stdlib-path`
flag is added to opt out of this default, making `/usr/lib/swift` the RPATH.

Add positive/negative tests.
`swift test` works on macOS for packages that import/use TensorFlow again (namely https://github.com/tensorflow/swift-apis).

Partially reverts https://github.com/apple/swift/pull/24787. Resolves [TF-797](https://bugs.swift.org/browse/TF-797).

dan-zheng committed:
Set default RPATH to toolchain stdlib instead of `/usr/lib/swift`.

Previously, the default RPATH was set to `/usr/lib/swift`.
This caused linker issues on macOS, as tensorflow-branch-specific modules like
TensorFlow and Python do not exist in `/usr/lib/swift`.

This patch defaults the `-toolchain-stdlib-rpath` flag to be true, so the
default RPATH is the toolchain standard library instead of `/usr/lib/swift`.

With this default, the linker issues are fixed. A `-no-toolchain-stdlib-path`
flag is added to opt out of this default, making `/usr/lib/swift` the RPATH.

Add positive/negative tests.
`swift test` works on macOS for packages that import/use TensorFlow again.

Partially reverts https://github.com/apple/swift/pull/24787.
Resolves TF-797.

bgogul reviewed (approved):
Thanks, Dan!

dan-zheng commented:
@swift-ci Please test tensorflow

brentdax commented:
Suggestion: Do most of this in master and just flip the default from `false` to `true` in TensorFlow.

dan-zheng commented:
> Suggestion: Do most of this in master and just flip the default from `false` to `true` in TensorFlow.

Done in https://github.com/apple/swift/pull/27207, please review.
If https://github.com/apple/swift/pull/27207 is merged, I'll then adapt `tensorflow` branch changes to minimize the diff.

dan-zheng committed:
Fix test/Driver/linker-rpath.swift, test implicit `-toolchain-stdlib-rpath`.

- Change existing driver invocations in Driver/linker-rpath to use explicit
  `-no-toolchain-stdlib-rpath` flag.
- Add tests for implicit `-toolchain-stdlib-rpath` flag.

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux CPU

ematejska commented:
Awesome!
